hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1505,6 +1883,35 @@ public boolean isAtomicRenameKey(String key) {
     return isKeyForDirectorySet(key, azureAtomicRenameDirSet);
   }
 
+  /**
+   * Provides a standard implementation of
+   * {@link org.apache.hadoop.fs.azurebfs.RenameAtomicityUtils.RedoRenameInvocation}.
+   */
+  RenameAtomicityUtils.RedoRenameInvocation getRedoRenameInvocation(final TracingContext tracingContext) {
+    return new RenameAtomicityUtils.RedoRenameInvocation() {
+      @Override
+      public void redo(final Path destination, final List<Path> sourcePaths,
+          final List<String> destinationSuffix)
+          throws AzureBlobFileSystemException {
+        for (int i = 0; i < sourcePaths.size(); i++) {
+          try {
+            final Path destinationPath;
+            if (destinationSuffix.get(i).isEmpty()) {
+              destinationPath = destination;
+            } else {
+              destinationPath = new Path(destination, destinationSuffix.get(i));
+            }
+            renameBlob(destinationPath, tracingContext, sourcePaths.get(i));
+          } catch (AbfsRestOperationException ex) {
+            if (ex.getStatusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
```

Comment:
What about other exception types, should we not throw the exception ?

Reply:
this is because certain blobs would have been renamed already in the initial crashed rename process. So that why handling 404 case and not other.

Make sense we should throw exception for other exception.

Reply:
very good catch there :).